### PR TITLE
feat(evaluation): commit the benchmark sample draw with its failure modes

### PR DESCRIPTION
**Pull request by ymohanty**
The 300-page Sarvam sample was drawn by a script living in a session scratchpad. Nobody else could reproduce it, and a scratchpad wipe would have taken both the sample and the code that produced it.

## The three failures this encodes

Each was found by running the draw for real on 9 August. Each produced a sample that looked correct in the selection log and was wrong in the manifest.

| # | Failure | Symptom |
|---|---|---|
| 1 | Allocation budgeted in pages, not documents | 301 pages covering **3 of 31** categories |
| 2 | Download order not interleaved | Page budget truncated the tail instead of thinning every class |
| 3 | No per-document page cap | One 22-page scan filled the COVID-19 cell, a category with **3 tickets** in a 36,909-ticket slice |

Each is now a named function with a regression test.

Also encoded: keys with a path separator are rejected. The evaluator derives the ticket from the file stem, so `AN063/E/2021/00001_complaint_….pdf` reads as ticket `00001` and joins to the wrong record.

## The lesson worth keeping

**The manifest is the artifact worth asserting on.** The selection log looked right in all three failures. Only the manifest showed the sample was not what was designed.

## Operational fact now recorded

Roughly **90% of the document corpus is in GLACIER**, where GetObject fails outright with `InvalidObjectState`. Restores must be requested and waited for, and the window has to outlast what the sample is for.

A 3-day window set on 9 August expired **13 August**, the day before the demo it was built for. Now extended to 31 August. This was not documented anywhere before today.

## Data safety

The manifest carries real ticket numbers and S3 keys, so it is citizen data by reference and must stay out of Git. `write_manifest` documents this and `data-check.yml` already refuses tracked files under `data/` and `outputs/`.

## Gate

    uv run ruff check .   -> All checks passed!
    uv run --extra serving --extra pipeline-core pytest -> 1292 passed, 8 skipped

## Note

This lands the draw as it exists today, which is **page-oriented for the Sarvam arm**. Moving to one shared sample across the standard pipeline, spam, dedup and routing will re-spine this on tickets. The pure functions here (allocation, interleaving, caps) are unit-agnostic and carry over; the S3/page projection is what changes.

Refs #53, #126, #127
---

## Merge note — Codex review round skipped

Merged without a further Codex round at the repository owner's explicit direction: *"the codex reviews keep returning issues each time without converging ... please then merge it idc"*.

Recording this plainly rather than implying a policy basis: `CONTRIBUTING.md` scopes the `codex-review-not-required` label to config-only branches and to Codex being out of review credits. **Neither applies here.** This is a deliberate override by the repository owner, labelled so the exception is visible in the record.

All findings raised on this PR were replied to in-thread and resolved. Any finding not fixed here is filed as a tracked issue rather than dropped.

### CODE REVIEW

**Comment by ymohanty** [comment]
@codex review

File: janasunani/evaluation/sarvam_sample_builder.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+    path.parent.mkdir(parents=True, exist_ok=True)
+    path.write_text(json.dumps(manifest, indent=2))
+    return path
+
+
+def draw_tickets(
+    rows: Iterable[tuple[str, str]],
+    *,
+    seed: int,
+) -> tuple[dict[str, list[str]], dict[str, int]]:
```
[the remark this answers is not in the document]
Comment (ymohanty) [reply janasunani/evaluation/sarvam_sample_builder.py:313]: Fixed in `fix(evaluation): give the sample draw an executable path and fix its arithmetic`.

This is the finding that mattered most, and it undermined the point of committing the module at all. The whole justification was "the sampler is in the repo now", and what was actually in the repo was the pieces. The operational draw stayed in an uncommitted scratchpad, and `is_ambiguous_key` was exercised only by its own unit test, never against a real key.

`build_sample` now runs the whole path: draw, look up, reject ambiguous keys, restore, interleave, cap, write the manifest, and delete anything fetched but not selected. Exposed as `janasunani-build-sarvam-sample`.

S3 sits behind a `DocumentSource` Protocol with the implementation in `sarvam_s3.py`. That is not ceremony: ~90% of this corpus is in GLACIER, so a test that needed a restore would never run. The fake source covers tickets with no document, ambiguous keys, and objects that never come back from archive.

Five tests on the composed path, including one asserting that every file left on disk is accounted for by the manifest, since a fetched-but-unselected file would let a later run read pages the manifest does not cover, and those pages get billed.

File: janasunani/evaluation/sarvam_sample_builder.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+    """Shuffle slice tickets per category under a fixed seed."""
+    by_category: dict[str, list[str]] = defaultdict(list)
+    for ticket, category in rows:
+        if category:
+            by_category[category].append(ticket)
+    rng = random.Random(seed)
+    for tickets in by_category.values():
+        rng.shuffle(tickets)
```
[the remark this answers is not in the document]
Comment (ymohanty) [reply janasunani/evaluation/sarvam_sample_builder.py:333]: Fixed in the same commit.

Verified before fixing: reversing your exact example changed the draw.

```
forward:  {"A": ["T3","T1","T2"], "B": ["U1","U2"]}
reversed: {"B": ["U2","U1"], "A": ["T1","T2","T3"]}
```

Both the per-category ticket lists and the category iteration order inherited the input ordering, and the rows come from a lake query with no `ORDER BY`. The manifest records a seed and claims the draw is reproducible from it, so this made that claim false rather than merely fragile.

Rows are canonicalised before the RNG is touched: categories sorted, tickets sorted, then shuffled. There is now a test asserting the draw is identical under reversed input, and that the category order is canonical too.

File: janasunani/evaluation/sarvam_sample_builder.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+    """
+    categories = sorted(counts)
+    alloc = {c: min(floor, counts[c]) for c in categories}
+    remaining = max(0, budget - sum(alloc.values()))
+    headroom = {c: counts[c] - alloc[c] for c in categories}
+    total_headroom = sum(headroom.values())
+    if remaining and total_headroom:
+        for category in categories:
+            extra = round(remaining * headroom[category] / total_headroom)
+            alloc[category] += min(extra, headroom[category])
```
[the remark this answers is not in the document]
Comment (ymohanty) [reply janasunani/evaluation/sarvam_sample_builder.py]: Fixed in the same commit.

Both of your examples reproduced exactly:

```
allocate(SLICE_COUNTS, budget=220, floor=3)  ->  221   (now 220)
allocate({A,B,C}, budget=5, floor=1)         ->    6   (now 5)
```

Independent rounding cannot preserve the total, and a function whose entire purpose is to respect a document budget overshooting it is not a rounding detail: the budget is what stops the draw exceeding the page target, which is what gets billed.

Now largest-remainder apportionment, with ties broken on category name so the outcome does not depend on input ordering either. Parametrised over five shapes including both of yours, asserting the total never exceeds the budget and does not leave it unspent when capacity exists.

**Comment by ymohanty** [comment]
@codex review

File: janasunani/evaluation/sarvam_s3.py [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+        self._client = client
+        self._region = region
+        self._profile = profile
+
+    @property
+    def client(self) -> Any:
+        if self._client is None:
+            import boto3
+
+            session = boto3.Session(profile_name=self._profile, region_name=self._region)
```
[the remark this answers is not in the document]
Comment (ymohanty) [reply janasunani/evaluation/sarvam_s3.py:100]: Fixed in d42ccee.

Verified against docs/DEPLOY.md before fixing: "Both share one IAM instance role (no static keys)" and "S3 access is via the instance role — no static keys on the boxes." A hardcoded `profile_name="Janasunani"` would indeed hit `ProfileNotFound` on those hosts.

`DEFAULT_PROFILE` is now `None`, so `boto3.Session(profile_name=None, ...)` falls through to the default credential chain (instance role on the deployed boxes). `S3DocumentSource(profile=...)` still accepts an explicit profile, and `main()` now exposes `--profile` for local use. New tests in `tests/test_sarvam_s3.py` cover both the unset default and an explicit override.

File: janasunani/evaluation/sarvam_sample_builder.py [hunk 5]
```diff
@@ -0,0 +1,497 @@
...
+    )
+    logger.info("selected {} documents across {} categories", len(chosen), len(counts))
+
+    pending = source.ensure_readable([doc["s3_key"] for doc in chosen])
+    if pending:
+        logger.warning("{} document(s) are still unavailable and will be skipped", len(pending))
+
+    out_dir.mkdir(parents=True, exist_ok=True)
+    available = [doc for doc in chosen if doc["s3_key"] not in pending]
+    ordered = interleave(available, counts)
```
[the remark this answers is not in the document]
Comment (ymohanty) [reply janasunani/evaluation/sarvam_sample_builder.py:486]: Fixed in d42ccee.

Verified before fixing: `build_sample` only `out_dir.mkdir(parents=True, exist_ok=True)`'d and the end-of-run cleanup only ever considered `fetched` (this run's documents), so a pre-existing file from an earlier draw was never touched — and `sarvam_evaluate.discover_pages` does enumerate every supported document under the directory with no reference to `sample_manifest.json`, so it would have been resubmitted and billed without appearing in the new manifest.

Added `prepare_out_dir`, called at the top of `build_sample`: raises `FileExistsError` on a non-empty `out_dir` by default, and only clears it when `clean_out_dir=True` (CLI: `--clean`). Tests assert both the rejection (with the leftover file left untouched) and the explicit clear.

File: janasunani/evaluation/sarvam_s3.py [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+            return set()
+
+        logger.info(
+            "restore requested ({}) for {} object(s)", self.restore_tier, len(pending)
+        )
+        waited = 0
+        while pending and waited < self.restore_timeout_seconds:
+            self._sleep(self.poll_seconds)
+            waited += self.poll_seconds
+            pending = {k for k in pending if self._restore_state(k) != "ready"}
```
[the remark this answers is not in the document]
Comment (ymohanty) [reply janasunani/evaluation/sarvam_s3.py:170]: Fixed in d42ccee.

Confirmed the fixed 1200s timeout applied regardless of `restore_tier` — `Standard`/`Bulk` were reachable via `--restore-tier` and `restore_timeout_seconds` had no tier awareness, so those requests would always time out and silently exclude every archived object (~90% of the corpus) from the manifest.

`restore_timeout_seconds` now defaults per tier via `DEFAULT_RESTORE_TIMEOUT_SECONDS` (Expedited 15 min, Standard 6h, Bulk 13h — padded above AWS's documented 3-5h/5-12h windows; an unrecognised tier falls back to the Bulk figure rather than the shortest one). An explicit value still overrides it, and `main()` now exposes `--restore-timeout-seconds`. New tests in `tests/test_sarvam_s3.py` check the tier floors, the override, the unknown-tier fallback, and that `ensure_readable` actually polls out a multi-hour window before giving up.

**Comment by ymohanty** [comment]
@codex review

File: janasunani/evaluation/sarvam_sample_builder.py [hunk 7]
```diff
@@ -0,0 +1,553 @@
...
+    billed for, since pages are the billed unit.
+    """
+    if path.suffix.lower() in IMAGE_SUFFIXES:
+        return 1
+    try:
+        from pdf2image import pdfinfo_from_path
+
+        return int(pdfinfo_from_path(str(path))["Pages"])
+    except Exception:
+        logger.warning("could not read a page count from {}; excluding it", path.name)
```
[the remark this answers is not in the document]
Comment (ymohanty) [reply janasunani/evaluation/sarvam_sample_builder.py:124]: Fixed in edaaf2d.

Verified before fixing: `pdf2image` is declared only under the `pipeline-core` and `ocr-deepseek` extras in `pyproject.toml`, and `janasunani-build-sarvam-sample` is a plain `[project.scripts]` entry with no extra binding. `page_count()`'s `except Exception` caught the resulting `ImportError` and returned 0, exactly like a corrupt PDF, so every fetched PDF would be excluded by `select_within_caps` and the CLI would exit 0 with an empty or truncated manifest.

`page_count()` now catches `ImportError` (and `PDFInfoNotInstalledError`, the same failure mode when Poppler itself is missing) separately and raises `PageCountBackendUnavailable` with the required invocation, rather than folding it into "unreadable document". `main()` catches it at the CLI boundary and reports it via `logger.error` + exit 1 instead of a raw traceback, matching the existing pattern in `sarvam_evaluate.py` for the same class of failure.

New tests: `test_page_count_raises_loudly_when_pdf2image_is_not_installed`, `test_image_pages_do_not_need_the_pdf_backend`, and `test_build_sample_raises_loudly_when_the_page_backend_is_missing` (composed path, asserts no manifest is written). All force the `ImportError` via `monkeypatch.setitem(sys.modules, "pdf2image", None)` so they're deterministic regardless of whether pdf2image happens to be installed in the test environment.

File: janasunani/evaluation/sarvam_sample_builder.py [hunk 8]
```diff
@@ -0,0 +1,553 @@
...
+    Ties break on category name so the result is reproducible rather than
+    dependent on the input's ordering.
+    """
+    categories = sorted(counts)
+    alloc = {c: min(floor, counts[c]) for c in categories}
+    remaining = max(0, budget - sum(alloc.values()))
+    headroom = {c: counts[c] - alloc[c] for c in categories}
+    total_headroom = sum(headroom.values())
+    if not remaining or not total_headroom:
+        return alloc
```
[the remark this answers is not in the document]
Comment (ymohanty) [reply janasunani/evaluation/sarvam_sample_builder.py:171]: Fixed in edaaf2d.

Verified your example before fixing: `allocate({"A": n, "B": n, "C": n}, budget=2, floor=1)` (three populated categories) returned an allocation totalling 3, over the budget of 2 — the floor stage was never checked against the budget before the largest-remainder apportionment ran. `choose_documents` then resolved the overshoot itself by stopping once `len(chosen) >= max_documents`, which truncates in largest-category order and silently drops whichever categories the floor was supposed to protect — exactly the tail-dropping failure this module already had a docstring warning about for a different reason.

`allocate()` now computes the floor allocation first and raises `ValueError` with a clear message (categories, floor, budget) if the floor total alone exceeds the budget, before any remainder apportionment runs. This is a reject-infeasible-input fix rather than a silent-degrade one: a caller with more categories than `budget / floor` allows needs to lower `--floor`, raise `--max-documents`, or narrow the category set, not get an answer that's already wrong.

New tests: `test_allocation_rejects_floors_that_exceed_the_budget` (your exact shape), `test_allocation_accepts_a_floor_that_exactly_fits_the_budget` (boundary case isn't over-rejected), and `test_choose_documents_propagates_the_infeasible_allocation_error` (the composed path doesn't swallow it and truncate instead). The five existing `test_allocation_never_exceeds_the_budget` parametrizations all stay under the new floor/budget check, so no prior case regresses.

**Comment by ymohanty** [comment]
@codex review

File: janasunani/evaluation/sarvam_sample_builder.py [hunk 9]
```diff
@@ -0,0 +1,630 @@
...
+            f"{out_dir} already has {len(leftover)} entr{'y' if len(leftover) == 1 else 'ies'} "
+            "in it. A prior draw's documents would be re-submitted alongside this one "
+            "without appearing in its manifest. Pass clean=True (CLI: --clean) to clear "
+            "it first, or point --out at an empty directory."
+        )
+    for entry in leftover:
+        if entry.is_dir():
+            shutil.rmtree(entry)
+        else:
+            entry.unlink()
```
[the remark this answers is not in the document]
Comment (ymohanty) [reply janasunani/evaluation/sarvam_sample_builder.py:441]: Confirmed real and taken seriously — filed as #246 and flagged there as a data-loss hazard rather than ordinary tech debt. A recursive delete of an unverified `--out` with no manifest, sentinel or path check is exactly the class this repository's operating rules say to treat as real.

Not fixed in this PR because #233 is merging without a further review round at the repository owner's direction, and because it is off the 14 August path: the Sarvam benchmark is blocked on account credit and `--clean` is invoked by no automated job. The fix wanted is refusing to clean a directory this tool does not recognise as its own staging area.
